File: PdfReader.cs
// PdfReader.cs
using System;
using System.IO;
using System.Text;
using System.Speech.Synthesis;
using iTextSharp.text.pdf;
using iTextSharp.text.pdf.parser;

class PdfReader
{
    static string ExtractText(string filePath, string pageRange)
    {
        StringBuilder sb = new StringBuilder();
        using (PdfReader reader = new PdfReader(filePath))
        {
            int totalPages = reader.NumberOfPages;
            int start = 1, end = totalPages;
            if (!string.IsNullOrEmpty(pageRange))
            {
                var parts = pageRange.Split('-');
                if (parts.Length == 2 && int.TryParse(parts[0], out int s) && int.TryParse(parts[1], out int e))
                {
                    start = Math.Max(1, s);
                    end = Math.Min(totalPages, e);
                }
            }
            for (int i = start; i <= end; i++)
            {
                sb.Append(PdfTextExtractor.GetTextFromPage(reader, i));
                sb.Append('\n');
            }
        }
        return sb.ToString();
    }

    static void Speak(string text, double speed = 1.0)
    {
        using (SpeechSynthesizer synth = new SpeechSynthesizer())
        {
            synth.Rate = (int)(speed * 2 - 2); // -10 to 10 range
            synth.Speak(text);
        }
    }

    static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: PdfReader.exe document.pdf [--page 1-5] [--speed 1.0] [--silent]");
            return;
        }
        string filePath = args[0];
        string pageRange = null;
        double speed = 1.0;
        bool silent = false;
        for (int i = 1; i < args.Length; i++)
        {
            if (args[i] == "--page" && i + 1 < args.Length)
                pageRange = args[++i];
            else if (args[i] == "--speed" && i + 1 < args.Length)
                double.TryParse(args[++i], out speed);
            else if (args[i] == "--silent")
                silent = true;
        }
        if (!File.Exists(filePath))
        {
            Console.WriteLine("File not found.");
            return;
        }
        string text = ExtractText(filePath, pageRange);
        if (!string.IsNullOrEmpty(text))
        {
            Console.WriteLine("Extracted text:\n" + (text.Length > 500 ? text.Substring(0, 500) + "..." : text));
            if (!silent)
                Speak(text, speed);
        }
        else
        {
            Console.WriteLine("No text extracted.");
        }
    }
}

File: pdf_reader.py
# pdf_reader.py
import sys
import argparse
import PyPDF2
import pyttsx3

def extract_text(pdf_path, page_range=None):
    text = ""
    with open(pdf_path, 'rb') as f:
        reader = PyPDF2.PdfReader(f)
        total_pages = len(reader.pages)
        if page_range:
            start, end = page_range
            start = max(0, start-1)
            end = min(total_pages, end)
        else:
            start, end = 0, total_pages
        for i in range(start, end):
            text += reader.pages[i].extract_text() + "\n"
    return text

def speak(text, speed=1.0, voice=None):
    engine = pyttsx3.init()
    if voice:
        voices = engine.getProperty('voices')
        for v in voices:
            if voice in v.name:
                engine.setProperty('voice', v.id)
                break
    engine.setProperty('rate', int(engine.getProperty('rate') * speed))
    engine.say(text)
    engine.runAndWait()

def main():
    parser = argparse.ArgumentParser(description='PDF Reader with TTS')
    parser.add_argument('file', help='PDF file path')
    parser.add_argument('--page', help='Page range: start-end (e.g., 1-5)')
    parser.add_argument('--speed', type=float, default=1.0, help='Speech speed (0.5-2.0)')
    parser.add_argument('--voice', help='Voice name (e.g., "Victoria" on macOS)')
    parser.add_argument('--silent', action='store_true', help='Only extract text, do not speak')
    args = parser.parse_args()

    page_range = None
    if args.page:
        parts = args.page.split('-')
        if len(parts) == 2:
            page_range = (int(parts[0]), int(parts[1]))
    text = extract_text(args.file, page_range)
    if text:
        print("Extracted text:\n", text[:500] + "..." if len(text) > 500 else text)
        if not args.silent:
            speak(text, args.speed, args.voice)
    else:
        print("No text extracted.")

if __name__ == "__main__":
    main()
